subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2335 @@
+	struct uvc_extension_unit_descriptor if_vc_xu;
+	struct uvc_output_terminal_descriptor if_vc_ot;
+
+	struct usb_if_descriptor if_vs;
+	struct uvc_stream_header_descriptor if_vs_header;
+	union uvc_stream_descriptor *if_vs_formats;
+	size_t if_vs_format_num;
+	struct usb_ep_descriptor if_vs_ep_fs;
+	struct usb_ep_descriptor if_vs_ep_hs;
+	struct uvc_color_descriptor if_vs_color;
+};
+
+struct uvc_probe {
+	uint16_t bmHint;
+	uint8_t bFormatIndex;
+	uint8_t bFrameIndex;
+	uint32_t dwFrameInterval;
+	uint16_t wKeyFrameRate;
+	uint16_t wPFrameRate;
+	uint16_t wCompQuality;
+	uint16_t wCompWindowSize;
+	uint16_t wDelay;
+	uint32_t dwMaxVideoFrameSize;
+	uint32_t dwMaxPayloadTransferSize;
+	uint32_t dwClockFrequency;
+	uint8_t bmFramingInfo;
+#define UVC_BMFRAMING_INFO_FID BIT(0)
+#define UVC_BMFRAMING_INFO_EOF BIT(1)
+#define UVC_BMFRAMING_INFO_EOS BIT(2)
+	uint8_t bPreferedVersion;
+	uint8_t bMinVersion;
+	uint8_t bMaxVersion;
+	uint8_t bUsage;
+	uint8_t bBitDepthLuma;
+	uint8_t bmSettings;
+	uint8_t bMaxNumberOfRefFramesPlus1;
+	uint16_t bmRateControlModes;
+	uint64_t bmLayoutPerStream;
+} __packed;
+
+struct uvc_payload_header {
+	uint8_t bHeaderLength;
+	uint8_t bmHeaderInfo;
+	uint32_t dwPresentationTime; /* optional */
+	uint32_t scrSourceClockSTC;  /* optional */
+	uint16_t scrSourceClockSOF;  /* optional */
+} __packed;
+
+/* Information specific to each VideoStreaming interface */
+struct uvc_stream {
+	const struct device *dev;
+	atomic_t state;
+	struct k_work work;
+	struct k_fifo fifo_in;
+	struct k_fifo fifo_out;
+	struct usbd_uvc_strm_desc *desc;
+	struct uvc_probe default_probe;
+	struct uvc_payload_header payload_header;
+	uint8_t format_id;
+	uint8_t frame_id;
+	const struct device *video_dev;
+	struct video_format video_fmt;
+	struct video_frmival video_frmival;
+	size_t vbuf_offset;
+};
+
+/* Global information */
+struct uvc_data {
+	atomic_t state;
+	struct usbd_class_data *c_data;
+	struct uvc_stream *streams;
+	struct usbd_uvc_desc *desc;
+	struct usb_desc_header **fs_desc;
+	struct usb_desc_header **hs_desc;
+	/* UVC error from latest request */
+	int err;
+};
+
+/* Specialized version of UDC net_buf metadata with extra fields */
+struct uvc_buf_info {
+	/* Regular UDC buf info so that it can be passed to USBD directly */
+	struct udc_buf_info udc;
+	/* Extra field at the end */
+	struct video_buffer *vbuf;
+	/* UVC stream this buffer belongs to */
+	struct uvc_stream *stream;
+} __packed;
+
+/* Mapping between UVC controls and Video controls */
+struct uvc_control_map {
+	uint8_t selector;
+	uint8_t size;
+	uint8_t bit;
+	enum uvc_op op;
+	uint32_t param;
+};
+
+struct uvc_guid_quirk {
+	uint32_t fourcc;
+	uint8_t guid[16];
+};
+
+typedef int uvc_control_fn_t(struct uvc_stream *strm, uint8_t selector, uint8_t request,
+			     struct net_buf *buf);
+
+NET_BUF_POOL_VAR_DEFINE(uvc_pool, DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) * 16,
+			512 * DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) * 16,
+			sizeof(struct uvc_buf_info), NULL);
```

Comment:
This is not a DCACHE-aware buffer pool. I will look in the next days how to provide a macro that also takes into account user data length argument.